sound/soc/sof/intel/hda-dai.c

```diff
@@ -217,6 +217,8 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
 		link_dev = hda_link_stream_assign(bus, substream);
 		if (!link_dev)
 			return -EBUSY;
+
+		snd_soc_dai_set_dma_data(dai, substream, (void *)link_dev);
```

Comment:
@plbossart is this correct? what happens if link_dev isnt null on line 215, then we'd never set the dma_data?

Reply:
well, that's the whole point: when is this dma_data supposed to be set?

I understood the code as doing the stream assignment once, so if it's null the dma_data had to be set earlier.
Also there are multiple return cases so it's possible with errors that this dma_data is never set, so I moved it just after the stream is allocated.

But to be honest I really don't understand the code. The pattern of testing for something before allocating it assumes a state machine...

Reply:
@plbossart, @RanderWang added the get_dma_data() where it is now based on the bug with multi-pipeline usecase: c7f1b858a69feaabb3c52bb4f1e111670a447220

Reply:
right, but the question is where should the set_dma_data() be done?
The way it's currently done is error prone, there are a set of return cases where this set_dma_data() would not be set, ever.

Reply:
@plbossart I think it should be done just outside the if (!link_dev) {} on line 223 instead

Reply:
then we'd set the set_dma_data multiple times, no very clear why?

Reply:
@plbossart you are correct. Sorry I dont know what I was reading this morning. Maybe its time to have breakfast!